text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1827 @@
+                               ┌───────────────────────┘
+                               ↓
+                     ┌──────────────────┐
+                     │ Pointee          │
+                     │ {runtime env, *} │
+                     └──────────────────┘
+                               │
+                            implies
+                               │
+                               ↓
+                     ┌──────────────────┐
+                     │ Value            │
+                     │ {runtime env, *} │
+                     └──────────────────┘
+```
+
+`Pointee` is still defined as being implemented for any type that can be used
+behind a pointer and may not be sized at all, this would be implemented for
+effectively every type except wasm's `externref` (or similar opaque types from
+some GPU targets). `Value` is defined as being implemented for any type that can
+be used as a value, which is all types, and also may not be sized at all.
+
+Earlier in this RFC, `extern type`s have previously been described as not being
+able to be used as a value, but it could instead be permitted to write functions
+which use extern types as values (e.g. such as taking an extern type as an argument),
+and instead rely on it being impossible to get a extern type that is not behind a
+pointer or a reference. This also implies that `ValueSized` types can be used as values,
+which would remain prohibited behind the `unsized_locals` and `unsized_fn_params`
+features until these are stabilised.
+
+With these changes to the RFC and possibility additional changes to the language, it
+could be possible to support wasm's `externref` and opaque types from some GPU targets.
+
+## Alignment
+[alignment]: #alignment
+
+There has been community interest in an [`Aligned` trait][rfc_aligned] and there
+are examples of `Aligned` traits being added in the ecosystem:
+
+- `rustc` has [its own `Aligned` trait][rustc_aligned] to support pointer tagging.
+- [`unsized-vec`][crate_unsized_vec] implements a `Vec` that depends on knowing
+  whether a type has an alignment or not.
+
+An `Aligned` trait hierarchy could be introduced alongside this proposal. It wouldn't
+viable to introduce `Aligned` within this hierarchy, as `dyn Trait` which is `ValueSized`
+would not be aligned, but some extern types could be `Aligned`, so there isn't an obvious
+place that an `Aligned` trait could be included in this hierarchy.
+
+## Custom DSTs
+[custom-dsts]: #custom-dsts
+
+Given the community interest in supporting custom DSTs in future (see
+[prior art][prior-art]), this RFC was written considering future-compatibility with
+custom DSTs in mind.
+
+There are various future changes to these traits which could be used to support
+custom DSTs on top of this RFC. None of these have been considered thoroughly, and are
+written here only to illustrate.
+
+- Allow `Pointee` to be implemented manually on user types, which would replace 
```

Comment:
```suggestion
- Allow `ptr::Pointee` to be implemented manually on user types, which would replace 
```

Presumably this refers to the existing `Pointee` and not the new one from this RFC, with the idea being that one can define a custom metadata for their custom DST?

Reply:
It does, clarified this.